migrations/src/main/scala/za/co/absa/enceladus/migrations/framework/migration/JsonMigration.scala

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright 2018-2019 ABSA Group Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.enceladus.migrations.framework.migration
+
+import org.apache.log4j.{LogManager, Logger}
+import za.co.absa.enceladus.migrations.framework.{MigrationUtils, ObjectIdTools}
+import za.co.absa.enceladus.migrations.framework.dao.DocumentDb
+
+import scala.collection.mutable
+
+/**
+  * A JsonMigration represents an entity that provides transformations for each document of every collection in a model
+  * when switching from one version of the model to another.
+  *
+  * All transformations are JSON string in, JSON string out.
+  *
+  * Only one transformation is possible per collection.
+  *
+  * In order to create a JSON migration you need to extend from this trait and provide all the requited transformations:
+  *
+  * {{{
+  *   object MigrationTo1 extends MigrationBase with JsonMigration {
+  *
+  *     transformJSON("collection1_name") (jsonIn => {
+  *       val jsonOut = collection1Transformations(jsonIn)
+  *       jsonOut
+  *     })
+  *
+  *     transformJSON("collection2_name") (jsonIn => {
+  *       val jsonOut = collection2Transformations(jsonIn)
+  *       jsonOut
+  *     })
+  *   }
+  * }}}
+  *
+  * If a source document is invalid for some reason and should be dropped, just return `InvalidDocument`.
+  *
+  */
+trait JsonMigration extends Migration {
+  type DocumentTransformer = String => String
+
+  private val log: Logger = LogManager.getLogger("JsonMigration")
+  val InvalidDocument = ""
+
+  private val transformers = new mutable.HashMap[String, DocumentTransformer]()
+
+  /**
+    * This function is used by derived classes to add transformations for affected collections.
+    * This is used for complex migrations that requite complex model version maps.
+    *
+    * @param collectionName A collection name to be migrated
+    * @param f              A transformation to applied to each document of the collection
+    *
+    */
+  def transformJSON(collectionName: String)(f: String => String): Unit = {
```

Comment:
This name sounds more like an alias to `applyTransformers`. Mabe `addTransformer`